File: mobile/lib/services/directions_service.dart
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import '../core/network/api_client.dart';
import '../core/network/api_error.dart';
import '../models/location.dart';
import '../models/traffic_condition.dart';

// ── Result types ───────────────────────────────────────────────────────────

/// Route data returned from a successful Directions API call.
class DirectionsRouteData {
  /// Normal travel time in seconds (no traffic).
  final int durationSeconds;

  /// Travel time with current traffic in seconds.
  final int durationInTrafficSeconds;

  /// Route distance in metres.
  final int distanceMetres;

  /// Calculated traffic status based on the congestion multiplier.
  final TrafficStatus status;

  /// Encoded polyline string (Google format) for drawing the real road path.
  /// Null when the response omits overview_polyline.
  final String? encodedPolyline;

  final DateTime timestamp;
  final DateTime eta;

  const DirectionsRouteData({
    required this.durationSeconds,
    required this.durationInTrafficSeconds,
    required this.distanceMetres,
    required this.status,
    required this.encodedPolyline,
    required this.timestamp,
    required this.eta,
  });

  String get etaFormatted => '${(durationInTrafficSeconds / 60).round()} mins';
  String get normalFormatted => '${(durationSeconds / 60).round()} mins';
  String get delayFormatted {
    final delaySec = durationInTrafficSeconds - durationSeconds;
    return delaySec <= 0 ? 'On time' : '+${(delaySec / 60).round()} mins';
  }
}

// ── Service ────────────────────────────────────────────────────────────────

/// Client-side Google Directions API integration.
///
/// The backend server handles ETA monitoring independently (see server/src/directionsService.ts).
/// This service is reserved for future client-side queries only — e.g. fetching the real
/// road polyline to draw on the map instead of a straight line.
///
/// SECURITY NOTE: The API key must NOT come from the bundled .env file. APK contents
/// are readable by anyone with apktool. When this service is activated, supply the key
/// via a server-side proxy, Android Keystore, or Firebase Remote Config — never hardcoded
/// or bundled. The key field below is intentionally left as a constructor parameter so
/// the caller controls the source.
///
/// Returns a typed record so the caller always handles both success and error branches.
class DirectionsService {
  static const String _tag = 'DirectionsService';
  static const String _directionsPath = '/maps/api/directions/json';

  /// [apiKey] must be injected by the caller from a secure source.
  /// Pass an empty string to get a [MissingApiKeyError] immediately.
  final String apiKey;

  const DirectionsService({required this.apiKey});

  Future<({DirectionsRouteData? data, ApiError? error})> fetchRoute({
    required AppLocation origin,
    required AppLocation destination,
  }) async {
    if (apiKey.trim().isEmpty) {
      debugPrint('[$_tag] apiKey is empty — skipping fetch');
      return (
        data: null,
        error: const MissingApiKeyError('Google Maps Directions'),
      );
    }

    try {
      final response = await ApiClient.googleMaps.get(
        _directionsPath,
        queryParameters: {
          'origin': '${origin.lat},${origin.long}',
          'destination': '${destination.lat},${destination.long}',
          'departure_time': 'now',
          'traffic_model': 'best_guess',
          'key': apiKey,
        },
      );

      return _parseResponse(response.data as Map<String, dynamic>);
    } on DioException catch (e) {
      if (e.response?.statusCode == 429) {
        final err = QuotaExceededError(_tag);
        debugPrint(err.logMessage);
        return (data: null, error: err);
      }
      final err = NetworkError(
        detail: '$_tag — ${e.message ?? "unknown"}',
        statusCode: e.response?.statusCode,
      );
      debugPrint(err.logMessage);
      return (data: null, error: err);
    } catch (e) {
      final err = NetworkError(detail: '$_tag — $e');
      debugPrint(err.logMessage);
      return (data: null, error: err);
    }
  }

  // ── Response parsing ─────────────────────────────────────────────────────

  ({DirectionsRouteData? data, ApiError? error}) _parseResponse(
      Map<String, dynamic> body) {
    final apiStatus = body['status'] as String?;

    if (apiStatus == 'REQUEST_DENIED') {
      final detail = body['error_message'] as String? ?? 'REQUEST_DENIED';
      final err = AuthDeniedError(service: _tag, detail: detail);
      debugPrint(err.logMessage);
      return (data: null, error: err);
    }

    if (apiStatus == 'OVER_DAILY_LIMIT' || apiStatus == 'OVER_QUERY_LIMIT') {
      final err = QuotaExceededError(_tag);
      debugPrint(err.logMessage);
      return (data: null, error: err);
    }

    if (apiStatus != 'OK') {
      final err = ParseError(field: 'status', detail: 'Unexpected: $apiStatus');
      debugPrint(err.logMessage);
      return (data: null, error: err);
    }

    final routes = body['routes'] as List<dynamic>?;
    if (routes == null || routes.isEmpty) {
      return (
        data: null,
        error: ParseError(field: 'routes', detail: 'Array is empty or missing'),
      );
    }

    final firstRoute = routes.first as Map<String, dynamic>;
    final legs = firstRoute['legs'] as List<dynamic>?;
    if (legs == null || legs.isEmpty) {
      return (
        data: null,
        error: ParseError(field: 'routes[0].legs', detail: 'Array is empty or missing'),
      );
    }

    final leg = legs.first as Map<String, dynamic>;

    final durationSec = _extractInt(leg, 'duration.value');
    final durationInTrafficSec =
        _extractInt(leg, 'duration_in_traffic.value') ?? durationSec;
    final distanceM = _extractInt(leg, 'distance.value');

    if (durationSec == null || distanceM == null) {
      return (
        data: null,
        error: ParseError(
          field: 'leg.duration / leg.distance',
          detail:
              'duration=$durationSec, durationInTraffic=$durationInTrafficSec, distance=$distanceM',
        ),
      );
    }

    final polyline =
        (firstRoute['overview_polyline'] as Map<String, dynamic>?)?['points']
            as String?;

    final multiplier = durationInTrafficSec! / durationSec;
    final now = DateTime.now();

    debugPrint(
      '[$_tag] Route fetched — '
      '${(durationInTrafficSec / 60).round()} min in traffic, '
      '${(distanceM / 1000).toStringAsFixed(1)} km, '
      'status=${_statusFromMultiplier(multiplier).name}'
    );

    return (
      data: DirectionsRouteData(
        durationSeconds: durationSec,
        durationInTrafficSeconds: durationInTrafficSec,
        distanceMetres: distanceM,
        status: _statusFromMultiplier(multiplier),
        encodedPolyline: polyline,
        timestamp: now,
        eta: now.add(Duration(seconds: durationInTrafficSec)),
      ),
      error: null,
    );
  }

  // ── Helpers ──────────────────────────────────────────────────────────────

  int? _extractInt(Map<String, dynamic> map, String dotPath) {
    final parts = dotPath.split('.');
    dynamic current = map;
    for (final part in parts) {
      if (current is Map<String, dynamic>) {
        current = current[part];
      } else {
        return null;
      }
    }
    return (current as num?)?.toInt();
  }

  TrafficStatus _statusFromMultiplier(double multiplier) {
    if (multiplier < 1.3) return TrafficStatus.calm;
    if (multiplier <= 1.8) return TrafficStatus.bookey;
    return TrafficStatus.ggs;
  }
}

File: server/src/directionsService.ts
import axios from 'axios';
import { serverConfig, hasGoogleMapsKey } from './config';

// ── Shared domain types ────────────────────────────────────────────────────

export interface Location {
  lat: number;
  long: number;
}

export type TrafficStatus = 'calm' | 'bookey' | "GG's";

export interface TrafficCondition {
  duration: number;           // normal travel time in seconds
  durationInTraffic: number;  // current travel time with traffic in seconds
  distance: number;           // route distance in metres
  status: TrafficStatus;
  timestamp: Date;
  eta: Date;
}

// ── Result union ───────────────────────────────────────────────────────────

type FetchSuccess = { ok: true; condition: TrafficCondition };

type FetchFailure = {
  ok: false;
  reason:
    | 'missing_key'       // key absent in config
    | 'auth_denied'       // key present but rejected by Google
    | 'quota_exceeded'    // billing quota hit
    | 'network_error'     // timeout or HTTP error
    | 'parse_error';      // unexpected response shape
  detail: string;
};

export type DirectionsResult = FetchSuccess | FetchFailure;

// ── Constants ──────────────────────────────────────────────────────────────

const DIRECTIONS_API_URL = 'https://maps.googleapis.com/maps/api/directions/json';
const REQUEST_TIMEOUT_MS = 8_000;

// ── Status helpers ─────────────────────────────────────────────────────────

export function statusFromMultiplier(multiplier: number): TrafficStatus {
  if (multiplier < 1.3) return 'calm';
  if (multiplier <= 1.8) return 'bookey';
  return "GG's";
}

// ── Live fetch ─────────────────────────────────────────────────────────────

export async function fetchDirectionsCondition(
  from: Location,
  to: Location
): Promise<DirectionsResult> {
  if (!hasGoogleMapsKey()) {
    console.error('[DirectionsService] No API key configured');
    return { ok: false, reason: 'missing_key', detail: 'GOOGLE_MAPS_DIRECTIONS_API_KEY is not set' };
  }

  try {
    const response = await axios.get(DIRECTIONS_API_URL, {
      timeout: REQUEST_TIMEOUT_MS,
      params: {
        origin: `${from.lat},${from.long}`,
        destination: `${to.lat},${to.long}`,
        departure_time: 'now',
        traffic_model: 'best_guess',
        key: serverConfig.googleMapsApiKey,
      },
    });

    return parseDirectionsResponse(response.data);
  } catch (error: unknown) {
    if (axios.isAxiosError(error)) {
      const httpStatus = error.response?.status;
      if (httpStatus === 429) {
        return {
          ok: false,
          reason: 'quota_exceeded',
          detail: `HTTP 429 from Directions API`,
        };
      }
      return {
        ok: false,
        reason: 'network_error',
        detail: `HTTP ${httpStatus ?? 'unknown'}: ${error.message}`,
      };
    }
    return { ok: false, reason: 'network_error', detail: String(error) };
  }
}

// ── Response parser ────────────────────────────────────────────────────────

function parseDirectionsResponse(body: unknown): DirectionsResult {
  if (typeof body !== 'object' || body === null) {
    return { ok: false, reason: 'parse_error', detail: 'Response body is not an object' };
  }

  const payload = body as Record<string, unknown>;
  const apiStatus = payload['status'];
  const errorMessage = payload['error_message'];

  if (apiStatus === 'REQUEST_DENIED') {
    return {
      ok: false,
      reason: 'auth_denied',
      detail: `REQUEST_DENIED — ${errorMessage ?? 'check API key and billing'}`,
    };
  }
  if (apiStatus === 'OVER_DAILY_LIMIT' || apiStatus === 'OVER_QUERY_LIMIT') {
    return { ok: false, reason: 'quota_exceeded', detail: `API status: ${apiStatus}` };
  }
  if (apiStatus !== 'OK') {
    return { ok: false, reason: 'parse_error', detail: `Unexpected API status: ${apiStatus}` };
  }

  const routes = payload['routes'];
  if (!Array.isArray(routes) || routes.length === 0) {
    return { ok: false, reason: 'parse_error', detail: 'routes[] is empty or missing' };
  }

  const legs = (routes[0] as Record<string, unknown>)['legs'];
  if (!Array.isArray(legs) || legs.length === 0) {
    return { ok: false, reason: 'parse_error', detail: 'routes[0].legs[] is empty or missing' };
  }

  const leg = legs[0] as Record<string, unknown>;

  const durationSec = ((leg['duration'] as Record<string, unknown>)?.['value']) as number | undefined;
  const durationInTrafficSec = ((leg['duration_in_traffic'] as Record<string, unknown>)?.['value']) as number | undefined;
  const distanceM = ((leg['distance'] as Record<string, unknown>)?.['value']) as number | undefined;

  if (durationSec === undefined || distanceM === undefined) {
    return {
      ok: false,
      reason: 'parse_error',
      detail: `Null numeric fields — duration=${durationSec}, distance=${distanceM}`,
    };
  }

  const effectiveDurationInTraffic = durationInTrafficSec ?? durationSec;
  const multiplier = effectiveDurationInTraffic / durationSec;
  const now = new Date();

  return {
    ok: true,
    condition: {
      duration: durationSec,
      durationInTraffic: effectiveDurationInTraffic,
      distance: distanceM,
      status: statusFromMultiplier(multiplier),
      timestamp: now,
      eta: new Date(now.getTime() + effectiveDurationInTraffic * 1_000),
    },
  };
}
